src/crossref.js

```diff
@@ -1,18 +1,272 @@
+                    const sourceItem = sourceItemsWithDOI[i];
+                    const newCitedItems = parsedItemReferences[i];
+                    if (newCitedItems.length > 0){
+                        const newCitations = newCitedItems.map((newItem) => new Citation({item: newItem, ocis: []}, sourceItem));
+                        sourceItem.addCitations(newCitations);
+                    }
+                }
+            })
+            progress.updateLine(
+                'done',
+                Wikicite.getString('wikicite.crossref.get-citations.done')
+            );
+        }
+        catch (error){
+            progress.updateLine(
+                'error',
+                Wikicite.getString('wikicite.crossref.get-citations.error-parsing-references')
+            );
+            debug(error);
+        }
+        finally {
+            progress.close();
+        }
+    }
+
+    /**
+     * Get a list of references from Crossref for an item with a certain DOI.
+     * Returned in JSON Crossref format.
+     * @param {string} doi - DOI for the item for which to get references.
+     * @returns {Promise<string[]>} list of references, or [] if none.
+     */
+    static async getReferences(doi) {
+        const JSONResponse = await Crossref.getCrossrefDOI(doi);
+        if (!JSONResponse || !JSONResponse.message.reference) {
+            return [];
+        }
+
+        return JSONResponse.message.reference;
+    }
+
+    /**
+     * Parse a list of references in JSON Crossref format.
+     * @param {string[]} crossrefReferences - Array of Crossref references to parse to Zotero items.
+     * @returns {Promise<Zotero.Item[]>} Zotero items parsed from references (where parsing is possible).
+     */
+    static async parseReferences(crossrefReferences){
+        if (!crossrefReferences.length){
+            debug("Item found in Crossref but doesn't contain any references");
+            return [];
+        }
+
+        const parsedReferences = await Promise.all(
+            crossrefReferences.map(async (reference) => await Crossref.parseReferenceItem(reference))
         );
-	}
+        return parsedReferences.filter(Boolean);
+    }
+
+    /**
+     * Parse a single item in JSON Crossref format to a Zotero Item.
+     * @param {string} crossrefItem - An reference item in JSON Crossref format.
+     * @returns {Promise<Zotero.Item | null>} Zotero item parsed from the Crossref reference, or null if parsing failed.
+     */
+    static async parseReferenceItem(crossrefItem){
+        let newItem = null;
+        if (crossrefItem.DOI){
+            newItem = await this.getItemFromIdentifier({DOI: crossrefItem.DOI});
+        }
+        else if(crossrefItem.isbn){
+            newItem = await this.getItemFromIdentifier({ISBN: crossrefItem.ISBN});
+        }
+        else{
+            newItem = this.parseItemFromCrossrefReference(crossrefItem);
+        }
+        return newItem;
+    }
```

Comment:
Merge these two methods (Annoyingly. the deletion in the diff means I can't make a suggestion).

```javascript
    /**
     * Parse a list of references in JSON Crossref format.
     * @param {string[]} crossrefReferences - Array of Crossref references to parse to Zotero items.
     * @returns {Promise<Zotero.Item[]>} Zotero items parsed from references (where parsing is possible).
     */
    static async parseReferences(crossrefReferences) {
        if (!crossrefReferences.length) {
            debug("Item found in Crossref but doesn't contain any references");
            return [];
        }

        const parsedReferences = await Promise.allSettled(
            crossrefReferences.map((crossrefItem) => {
                if (crossrefItem.DOI)
                    return this.getItemFromIdentifier({ DOI: crossrefItem.DOI });

                if (crossrefItem.isbn)
                    return this.getItemFromIdentifier({ ISBN: crossrefItem.ISBN });

                return this.parseItemFromCrossrefReference(crossrefItem);
            })
        );
        return parsedReferences
            .map((reference) => reference.value || null)
            .filter(Boolean);
    }
```

I've implemented the `Promise.allSettled` fix, here — please discard if you want to fail if any individual parse fails. Given all these functions return promises, I've assumed you'll user `Promise.reject` for errors, rather than returning `Promise.resolve(null)`.

Reply:
Yeah, this point I'll have to think about more - whether we want to add references to an item if that list of references is incomplete. But then, we already silently discard the unstructured text references anyway so I don't think it's a big deal.

Reply:
I don't like it but best effort (allowing incomplete reference lists) is a reasonable approach for now, especially given the discussion around #25.